doc/krillc.1
.TH "krillc" "1" "NLnet Labs"
.SH NAME
krillc - Command line client for the Krill RPKI daemon
.SH SYNOPSIS
krillc [SUBCOMMAND]

krillc help [SUBCOMMAND]

krill -h, --help

krill -V, --version
.SH DESCRIPTION
Krill is a free, open source RPKI Certificate Authority that lets you run
delegated RPKI under one or multiple Regional Internet Registries (RIRs). 
Through its built-in publication server, Krill can publish Route Origin
Authorizations (ROAs) on your own servers or with a third party.

This manual page documents the krillc command line interface that interacts with
a krill daemon.

For more information on the extensive set of subcommands
supported by krillc please consult the online documentation at:
    https://krill.docs.nlnetlabs.nl/en/stable/cli.html

The command line client also has built-in help which can be accessed like so:
    krillc help [SUBCOMMAND]

.SH SEE ALSO
krill(1)


Changelog.md
# Change Log

Please see [here](https://github.com/NLnetLabs/krill/projects?query=is%3Aopen+sort%3Aname-asc)
for planned releases.

## 0.7.5 'Time's Up'

This release is not intended for public use, please upgrade to the latest version - which is
Krill 0.9.3 'Thundering Herd' - if you can. We introduced this version as a quick fix to
deal with http client timeouts for publishing CAs.

## 0.7.4 'Multipass!'

There is no need to upgrade to this version. It was created only so that you can continue
to compile Krill locally using the latest Rust compiler.

As it turns out the use of many asynchronous calls, the cool stuff which make Krill thread safe,
cause the compiler to do quite a bit of work in a process called 'Monomorphization'. The latest
compiler version will go on strike as a result, unless we instruct it beforehand that more work
is coming its way.

## 0.7.3 'Slow Food'

This release fixes an issue where the BGP Ris Dump files were reloaded and checked too
frequently causing high CPU and bandwidth usage.

## 0.7.2 'Small Bites'

This release fixes an issue where BGP RIS Dump files that were not properly retrieved would
cause a thread to choke. As this can lead to lock poisoning this type of event could cause
other Krill processes to stop functioning properly. All users of Krill 0.7.0 and 0.7.1 are
advised to upgrade.

In addition to this German translations have been added to the UI.

## 0.7.1 'Sobremesa'

This release fixes the ROA migration introduced in 0.7.0. We identified an issue where the
clean up of ROAs would fail because Krill tried adding explicit forms of ROAs - with max
length set - before removing the implicit definitions.

## 0.7.0 'Escondidinho de Lagosta'

This release brings significant improvements aimed at maintaining your ROAs. For now, Krill
will download aggregated BGP dumps from the RIPE NCC Routing Information Service (*) and
analyse how your ROAs affect announcements seen for your resources. In future we will extend
this system, so that it can use near-real-time data, or even a local feed with your own BGP
information instead. 

For these changes to work well we needed to do some work on cleaning up existing ROAs. Until
now Krill has allowed the creation of essentially duplicate, or nonsensical ROAs, such as:
* ROAs for an ASN and prefix with and without an explicit max length matching the prefix 
* ROAs for a prefix and ASN which were already permitted by another ROA.  

On upgrade Krill will clean up such redundant authorizations for ROAs. For example if the
following authorizations would exist:

 192.168.0.0/16      => 64496
 192.168.0.0/24      => 64496
 192.168.0.0/16-24   => 64496
 
Then only this last authorization needs to be kept, the first two are also covered by it.

Before this release it was also possible to have the same authorization with, and without, using
an explicit max length. For example:

 192.168.0.0/16      => 64496
 192.168.0.0/16-16   => 64496

Now Krill will *always* use an explicit max length in the definitions. Note however, that it is
still best practice to use the same max length as the announced prefix length, so Krill will just
set this by default if it is not specified.

*: https://www.ripe.net/analyse/internet-measurements/routing-information-service-ris

## 0.6.3 'Play it again, Sam'

This release addresses an issue where users with a CA that has delegated children, which in turn
had performed a key roll over in the past, could not upgrade to Release 0.6.2.

Users who already successfully upgraded to Release 0.6.2 do not need to upgrade urgently. This
release includes a number of fixes for minor issues, which will also be included in the 0.7.0
Release which is due in 2-4 weeks:
* `krillc issues` fails with `Error: Unknown API method` (#248)
* `krillc parents` help text refers incorrectly to publisher request instead of child request (#251)
* Normalize request/response `krillc help` texts (#252)
* `krillc` incorrectly reports XML as a supported output format (#253)
* Inconsistent use of "cas" in `krillc bulk` subcommand summary text (#254)
* Be consistent when referring to ending with a / (#255)

## 0.6.2 'That was even faster!'

So, as it turns out.. the code used to determine the age of snapshot files used in the previous
release was not safe on all platforms. This release fixes this!

Users who upgraded to 0.6.1 and see messages like: "Creation time is not available on this
platform currently" in their logs, please upgrade!

## 0.6.1 'That was fast!'

This release fixes an issue where the Krill Repository Server deleted RRDP snapshot files as soon
as a new notification file was published. This leads to issues in case a cached notification file
is served to validators. 

Users who use Krill as their own Repository Server are advised to upgrade.

Users who publish at a repository provided to them by a third party (e.g. nic.br) can safely skip
this release.  

## 0.6.0 'Go with the Flow'

The most visible change in this release is that the embedded Lagosta UI now includes French, Greek
and Spanish translations. But, the vast majority of the work went into making Krill use asynchronous
code.

We migrated from actix-web to Hyper. Hyper is a fast, safe and fully asynchronous web framework
which has a lot of momentum behind it. This change also meant that we needed to ensure that
Krill itself uses safe asynchronous code whenever it connects to a remote system, like a parent
or repository, or in case of the CLI the Krill API itself.

In addition to this we improved the history API to ensure that Krill will no longer use an
excessive amount of history in cases where a CA has a long history. The API is still subject to
change, and therefore we will only document this in future. In the meantime however, the CLI
may be used to show the history of your CA.

Lagosta:
* Now includes French, Greek and Spanish translations
* Minor improvements in functionality

Krill back-end:
* Migrated from actix-web to hyper.

src/client/krillc.rs
use std::io;
use std::time::Duration;
use bytes::Bytes;
use reqwest::{Client, Response, StatusCode};
use reqwest::header::{HeaderMap, HeaderValue, USER_AGENT};
use rpki::uri;
use crate::client::data::{
    ApiResponse,
    PublisherDetails,
    PublisherList,
    ReportError
};
use crate::client::options::{
    Options,
    Command,
    PublishersCommand
};
use crate::util::file;

/// Command line tool for Krill admin tasks
pub struct KrillClient {
    server: uri::Http,
    token: String
}

impl KrillClient {

    /// Delegates the options to be processed, and reports the response
    /// back to the user. Note that error reporting is handled by CLI.
    pub fn report(options: Options) -> Result<(), Error> {
        let format = options.format.clone();
        let res = Self::process(options)?;

        if let Some(string) = res.report(format)? {
            println!("{}", string)
        }
        Ok(())
    }

    /// Processes the options, and returns a response ready for formatting.
    /// Note that this function is public to help integration testing the API
    /// and client.
    pub fn process(options: Options) -> Result<ApiResponse, Error> {
        let client = KrillClient {
            server: options.server,
            token:  options.token
        };

        match options.command {
            Command::Health => client.health(),
            Command::Publishers(cmd) => client.publishers(cmd),
            Command::NotSet => Err(Error::MissingCommand)
        }
    }

    fn health(&self) -> Result<ApiResponse, Error> {
        self.get("api/v1/health")?;
        Ok(ApiResponse::Health)
    }

    fn publishers(
        &self,
        command: PublishersCommand,
    ) -> Result<ApiResponse, Error> {
        match command {
            PublishersCommand::List => {
                let res = self.get("api/v1/publishers")?;
                let list: PublisherList = serde_json::from_str(&res)?;
                Ok(ApiResponse::PublisherList(list))
            },
            PublishersCommand::Add(path, handle_opt) => {
                let xml_bytes = file::read(&path)?;
                let uri = match handle_opt {
                    None => "api/v1/publishers".to_string(),
                    Some(handle) => format!("api/v1/publishers/{}", handle)
                };
                match self.post(uri.as_str(), xml_bytes)? {
                    Some(body) => {
                        if body.is_empty() {
                            Ok(ApiResponse::Empty)
                        } else {
                            Ok(ApiResponse::GenericBody(body))
                        }
                    },
                    None => Ok(ApiResponse::Empty)
                }
            },
            PublishersCommand::Remove(handle) => {
                let uri = format!("api/v1/publishers/{}", handle);
                self.delete(&uri)?;
                Ok(ApiResponse::Empty)
            },
            PublishersCommand::Details(handle) => {
                let uri = format!("api/v1/publishers/{}", handle);
                let res = self.get(uri.as_str())?;
                let details: PublisherDetails = serde_json::from_str(&res)?;
                Ok(ApiResponse::PublisherDetails(details))
            },
            PublishersCommand::RepositoryResponseXml(handle, file_opt) => {
                let uri = format!("api/v1/publishers/{}/response.xml", handle);
                let xml = self.get(uri.as_str())?;
                match file_opt {
                    Some(path) => {
                        file::save(&Bytes::from(xml), &path)?;
                        Ok(ApiResponse::Empty)
                    },
                    None => {
                        Ok(ApiResponse::GenericBody(xml))
                    }
                }
            },
            PublishersCommand::IdCert(handle, file) => {
                let uri = format!("api/v1/publishers/{}", handle);
                let res = self.get(uri.as_str())?;
                let details: PublisherDetails = serde_json::from_str(&res)?;
                match details.identity_cert() {
                    Some(cert) => {
                        let bytes = cert.to_bytes();
                        file::save(&bytes, &file)?;
                        Ok(ApiResponse::Empty)
                    },
                    None => Err(Error::NoIdCert)
                }

            }
        }
    }

    fn headers(&self) -> HeaderMap {
        let mut headers = HeaderMap::new();
        headers.insert(
            USER_AGENT,
            HeaderValue::from_str("krillc").unwrap()
        );
        headers.insert(
            "Authorization",
            HeaderValue::from_str(&format!("Bearer {}", &self.token)).unwrap()
        );
        headers
    }

    fn client(&self) -> Result<Client, Error> {
        Client::builder()
            .gzip(true)
            .timeout(Duration::from_secs(30))
            .build()
            .map_err(|e| Error::RequestError(e))
    }

    fn post(&self, rel: &str, bytes: Bytes) -> Result<Option<String>, Error> {
        let headers = self.headers();

        let uri = format!("{}{}", &self.server.to_string(), rel);
        let mut res = self.client()?.post(&uri)
            .headers(headers)
            .body(bytes.to_vec())
            .send()?;

        match res.status() {
            StatusCode::OK => {
                Ok(res.text().ok())
            },
            status => {
                match res.text() {
                    Ok(body) => {
                        if body.is_empty() {
                            Err(Error::BadStatus(status))
                        } else {
                            Err(Error::ErrorWithBody(body))
                        }
                    },
                    _ => Err(Error::BadStatus(status))
                }
            }
        }
    }

    /// Sends a get request to the server, including the token for
    /// authorization.
    /// Note that the server uri ends with a '/', so leave out the '/'
    /// from the start of the rel_path when calling this function.
    fn get(
        &self,
        rel_path: &str
    ) -> Result<String, Error> {
        let mut res = self.get_generic(rel_path)?;
        let txt = res.text()?;
        Ok(txt)
    }

    /// Sends a get request to the server, including the token for
    /// authorization.
    /// Note that the server uri ends with a '/', so leave out the '/'
    /// from the start of the rel_path when calling this function.
    #[allow(dead_code)]
    fn get_binary(
        &self,
        rel_path: &str
    ) -> Result<Bytes, Error> {
        let mut res = self.get_generic(rel_path)?;
        let mut bytes: Vec<u8> = vec![];
        res.copy_to(&mut bytes)?;
        Ok(Bytes::from(bytes))
    }

    fn get_generic(
        &self,
        rel_path: &str
    ) -> Result<Response, Error> {
        let uri = format!("{}{}", &self.server.to_string(), rel_path);
        let mut res = self.client()?
            .get(&uri)
            .headers(self.headers())
            .send()?;

        match res.status() {
            StatusCode::OK => {
                Ok(res)
            },
            status => {
                match res.text() {
                    Ok(body) => {
                        if body.is_empty() {
                            Err(Error::BadStatus(status))
                        } else {
                            Err(Error::ErrorWithBody(body))
                        }
                    },
                    _ => Err(Error::BadStatus(status))
                }
            }
        }
    }

    fn delete(
        &self,
        rel: &str
    ) -> Result<(), Error> {
        let uri = format!("{}{}", &self.server.to_string(), rel);
        let res = self.client()?
            .delete(&uri)
            .headers(self.headers())
            .send()?;

        match res.status() {
            StatusCode::OK => Ok(()),
            status         => Err(Error::BadStatus(status))
        }
    }

}

//------------ Error ---------------------------------------------------------

#[derive(Debug, Display)]
pub enum Error {
    #[display(fmt="No valid command given, see --help")]
    MissingCommand,

    #[display(fmt="Server is not available.")]
    ServerDown,

    #[display(fmt="Request Error: {}", _0)]
    RequestError(reqwest::Error),

    #[display(fmt="Received bad status: {}", _0)]
    BadStatus(StatusCode),

    #[display(fmt="{}", _0)]
    ErrorWithBody(String),

    #[display(fmt="Received invalid json response: {}", _0)]
    JsonError(serde_json::Error),

    #[display(fmt="{}", _0)]
    ReportError(ReportError),

    #[display(fmt="Can't read file: {}", _0)]
    IoError(io::Error),

    #[display(fmt="There is no known IdCert for this publisher")]
    NoIdCert
}

impl From<reqwest::Error> for Error {
    fn from(e: reqwest::Error) -> Self {
        Error::RequestError(e)
    }
}

impl From<serde_json::Error> for Error {
    fn from(e: serde_json::Error) -> Self {
        Error::JsonError(e)
    }
}

impl From<io::Error> for Error {
    fn from(e: io::Error) -> Self {
        Error::IoError(e)
    }
}

impl From<ReportError> for Error {
    fn from(e: ReportError) -> Self {
        Error::ReportError(e)
    }
}


// Note: this is all tested through integration tests ('tests' folder).